File: src/ai/heuristics.js
(function(global) {
    const Config = (typeof require !== 'undefined') ? require('./config') : global.SnakeAI.Config;
    const floodFill = (typeof require !== 'undefined') ? require('./floodfill') : global.SnakeAI.floodFill;
    const computeVoronoi = (typeof require !== 'undefined') ? require('./voronoi') : global.SnakeAI.computeVoronoi;

    function manhattan(p1, p2) {
        return Math.abs(p1.x - p2.x) + Math.abs(p1.y - p2.y);
    }

    function evaluate(grid, state, debug = false) {
        const me = state.me; 
        const enemy = state.enemy;
        const breakdown = {}; 

        if (me.health <= 0) return debug ? { total: Config.SCORES.LOSS, reason: "Dead" } : Config.SCORES.LOSS;
        if (enemy.health <= 0) return debug ? { total: Config.SCORES.WIN, reason: "Enemy Dead" } : Config.SCORES.WIN;

        let score = 0;
        const occupancy = (me.body.length + enemy.body.length) / (grid.width * grid.height);
        const denseTailRace = (
            me.body.length >= 20 &&
            enemy.body.length >= 20 &&
            occupancy >= Config.DENSE_TAIL_RACE_OCCUPANCY
        );

        // 1. Length
        const lengthScore = me.body.length * Config.SCORES.LENGTH;
        score += lengthScore;
        if (debug) breakdown.length = lengthScore;

        const myHead = me.body[0];
        const enemyHead = enemy.body[0];
        
        // Handle Tail Safety for the simulation
        let tailIsSafe = false;
        let originalTailVal = 0;
        const tail = me.body[me.body.length - 1];
        if (tail && me.health < 100) {
            tailIsSafe = true;
            originalTailVal = grid.get(tail.x, tail.y);
            grid.set(tail.x, tail.y, 0); 
        }

        // 2. Voronoi Territory Control
        const voronoi = computeVoronoi(grid, myHead, enemyHead);
        
        if (tailIsSafe) grid.set(tail.x, tail.y, originalTailVal);

        // A. Territory Score
        const territoryScore = (voronoi.myCount - voronoi.enemyCount) * Config.SCORES.TERRITORY_CONTROL;
        score += territoryScore;

        let iAmInDeathTrap = false; 

        if (voronoi.myCount < me.body.length) {
            
            const ffResult = floodFill(grid, myHead.x, myHead.y, me.body.length + 2, me.body);
            
            const physicalSpace = ffResult.count;
            const adjustedEscapeTime = ffResult.minTurnsToClear + (ffResult.hasFood ? 1 : 0);
            const futureLength = me.body.length + (ffResult.hasFood ? 1 : 0);

            const isTrapped = (physicalSpace < futureLength) && (physicalSpace < adjustedEscapeTime);

            if (isTrapped) {
                iAmInDeathTrap = true;
                const trapDeathPenalty = denseTailRace
                    ? (Config.SCORES.TRAP_DANGER * 0.001)
                    : Config.SCORES.TRAP_DANGER;
                score += trapDeathPenalty;
                if (debug) breakdown.trapDeath = trapDeathPenalty;
            } else if (physicalSpace >= futureLength) {
                const pressurePenalty = Config.SCORES.STRATEGIC_SQUEEZE; 
                score += pressurePenalty;
                if (debug) breakdown.pressure = pressurePenalty;
            }
        } 
        else if (voronoi.myCount < grid.width * grid.height * 0.2) {
             score += Config.SCORES.TIGHT_SPOT;
             if (debug) breakdown.tight = Config.SCORES.TIGHT_SPOT;
        }

        // 3. Enemy Trap Detection
        if (!iAmInDeathTrap && voronoi.enemyCount < enemy.body.length) {
            const enemyHead = enemy.body[0];
            const enFF = floodFill(grid, enemyHead.x, enemyHead.y, enemy.body.length + 2, enemy.body);
            
            const enSpace = enFF.count;
            const enTimeToEscape = enFF.minTurnsToClear;
            const enFutureLen = enemy.body.length + (enFF.hasFood ? 1 : 0);

            if (enSpace < enFutureLen && enSpace < enTimeToEscape) {
                const enemyTrappedBonus = denseTailRace
                    ? (Config.SCORES.ENEMY_TRAPPED * 0.001)
                    : Config.SCORES.ENEMY_TRAPPED;
                score += enemyTrappedBonus;
                if (debug) breakdown.killThreat = enemyTrappedBonus;
            }
        }

        // 4. Tactical Kill Pressure
        const distToOpp = manhattan(myHead, enemyHead);
        if (distToOpp === 1 && me.body.length > enemy.body.length) {
            const killScore = Config.SCORES.KILL_PRESSURE;
            score += killScore;
            if (debug) breakdown.killPressure = killScore;
        }

        // 4. Food
        let foodScore = 0;
        if (state.food && state.food.length > 0) {
            let closestDist = 9999;
            if (state.distMap) {
                const headIdx = myHead.y * grid.width + myHead.x;
                closestDist = state.distMap[headIdx];
            } else {
                for(const f of state.food) {
                    const d = manhattan(myHead, f);
                    if (d < closestDist) closestDist = d;
                }
            }

            if (closestDist > me.health) {
                return debug ? { total: Config.SCORES.LOSS, reason: "Starvation" } : Config.SCORES.LOSS;
            }

            let panicValue = 0;
            const buffer = me.health - closestDist;
            
            if (buffer > 0) {
                panicValue = Config.SCORES.FOOD.INTENSITY * 
                    Math.pow(Config.SCORES.FOOD.THRESHOLD / (buffer + 1), Config.SCORES.FOOD.EXPONENT);
            } else {
                panicValue = Config.SCORES.FOOD.INTENSITY * 100;
            }
            foodScore -= closestDist * panicValue;
        }
        score += foodScore;
        if (debug) breakdown.foodCurve = Math.floor(foodScore);

        // 5. Aggression
        let aggroScore = 0;
        if (me.body.length > enemy.body.length + 1) {
            aggroScore = -(distToOpp * Config.SCORES.AGGRESSION);
        }
        score += aggroScore;
        if (debug) breakdown.aggro = aggroScore;

        if (debug) {
            breakdown.total = score;
            return breakdown;
        }
        return score;
    }

    if (typeof module !== 'undefined' && module.exports) {
        module.exports = evaluate;
    } else {
        global.SnakeAI = global.SnakeAI || {};
        global.SnakeAI.evaluate = evaluate;
    }
})(this);

File: rust/crates/snake-ai/src/heuristics.rs
use snake_domain::Point;

use crate::{
    config::AiConfig,
    floodfill::flood_fill,
    grid::Grid,
    model::{AgentState, SearchBuffers},
    voronoi::compute_voronoi,
};

#[inline]
fn manhattan(a: Point, b: Point) -> i32 {
    (a.x - b.x).abs() + (a.y - b.y).abs()
}

// WRAPPER FOR PROFILING
pub fn evaluate(
    grid: &mut Grid,
    me: &AgentState,
    enemy: &AgentState,
    food: &[Point],
    dist_map: Option<&[i16]>,
    cfg: &AiConfig,
    buffers: &mut SearchBuffers,
) -> f64 {
    let start = std::time::Instant::now();
    let res = evaluate_inner(grid, me, enemy, food, dist_map, cfg, buffers);
    crate::PERF_STATS.with(|s| {
        let mut st = s.borrow_mut();
        st.eval_calls += 1;
        st.eval_duration += start.elapsed();
    });
    res
}

fn evaluate_inner(
    grid: &mut Grid,
    me: &AgentState,
    enemy: &AgentState,
    food: &[Point],
    dist_map: Option<&[i16]>,
    cfg: &AiConfig,
    buffers: &mut SearchBuffers,
) -> f64 {
    if me.health <= 0 || me.body.is_empty() {
        return cfg.scores.loss;
    }
    if enemy.health <= 0 || enemy.body.is_empty() {
        return cfg.scores.win;
    }

    let mut score = 0.0;
    let occupancy = (me.body.len() + enemy.body.len()) as f64 / (grid.width * grid.height) as f64;
    let dense_tail_race = me.body.len() >= 20 && enemy.body.len() >= 20 && occupancy >= cfg.dense_tail_race_occupancy;

    score += me.body.len() as f64 * cfg.scores.length;

    let my_head = me.body[0];
    let enemy_head = enemy.body[0];

    let mut tail_is_safe = false;
    let mut original_tail_val = 0i8;
    if let Some(tail) = me.body.last().copied() {
        if me.health < 100 {
            tail_is_safe = true;
            original_tail_val = grid.get(tail.x, tail.y);
            grid.set(tail.x, tail.y, 0);
        }
    }

    let voronoi = compute_voronoi(grid, my_head, enemy_head, buffers);
    if tail_is_safe {
        let tail = me.body[me.body.len() - 1];
        grid.set(tail.x, tail.y, original_tail_val);
    }

    score += (voronoi.my_count - voronoi.enemy_count) as f64 * cfg.scores.territory_control;

    let mut i_am_in_death_trap = false;
    if voronoi.my_count < me.body.len() as i32 {
        let ff = flood_fill(grid, my_head.x, my_head.y, me.body.len() as i32 + 2, Some(&me.body), buffers);
        let physical_space = ff.count;
        let adjusted_escape_time = ff.min_turns_to_clear + if ff.has_food { 1 } else { 0 };
        let future_len = me.body.len() as i32 + if ff.has_food { 1 } else { 0 };

        let trapped = physical_space < future_len && physical_space < adjusted_escape_time;
        if trapped {
            i_am_in_death_trap = true;
            score += if dense_tail_race {
                cfg.scores.trap_danger * 0.001
            } else {
                cfg.scores.trap_danger
            };
        } else if physical_space >= future_len {
            score += cfg.scores.strategic_squeeze;
        }
    } else if voronoi.my_count < ((grid.width * grid.height) as f64 * 0.2) as i32 {
        score += cfg.scores.tight_spot;
    }

    if !i_am_in_death_trap && voronoi.enemy_count < enemy.body.len() as i32 {
        let en_head = enemy.body[0];
        let en_ff = flood_fill(grid, en_head.x, en_head.y, enemy.body.len() as i32 + 2, Some(&enemy.body), buffers);
        let en_space = en_ff.count;
        let en_escape = en_ff.min_turns_to_clear;
        let en_future_len = enemy.body.len() as i32 + if en_ff.has_food { 1 } else { 0 };
        if en_space < en_future_len && en_space < en_escape {
            score += if dense_tail_race {
                cfg.scores.enemy_trapped * 0.001
            } else {
                cfg.scores.enemy_trapped
            };
        }
    }

    let dist_to_opp = manhattan(my_head, enemy_head);
    if dist_to_opp == 1 && me.body.len() > enemy.body.len() {
        score += cfg.scores.kill_pressure;
    }

    if !food.is_empty() {
        let closest_dist = if let Some(map) = dist_map {
            map[(my_head.y * grid.width + my_head.x) as usize] as i32
        } else {
            food.iter().map(|f| manhattan(my_head, *f)).min().unwrap_or(9999)
        };

        if closest_dist > me.health {
            return cfg.scores.loss;
        }

        let buffer = me.health - closest_dist;
        let panic_value = if buffer > 0 {
            cfg.scores.food.intensity * (cfg.scores.food.threshold / (buffer as f64 + 1.0)).powf(cfg.scores.food.exponent)
        } else {
            cfg.scores.food.intensity * 100.0
        };
        score -= closest_dist as f64 * panic_value;
    }

    if me.body.len() > enemy.body.len() + 1 {
        score -= dist_to_opp as f64 * cfg.scores.aggression;
    }

    score
}
